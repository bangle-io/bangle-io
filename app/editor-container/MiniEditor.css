.B-editor-container_mini-editor {
  background-color: var(--BV-window-popup-editor-bg-color);
  color: var(--BV-window-popup-editor-color);

  /* setting width and height are important otherwise
      tippy confuses and causes issue with scroll
    */
  width: 400px;
  height: min(50vh, 600px);
  z-index: var(--BV-window-popup-editor-z-index);
  border: 1px solid var(--BV-window-border-color-0);
  box-shadow: rgb(0 0 0 / 30%) 0 4px 28px;
  border-radius: 0.25rem;
  position: fixed;
  display: flex;
  flex-direction: column;
  bottom: 0;
  right: 30px;
}
