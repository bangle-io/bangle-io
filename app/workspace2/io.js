import * as idb from 'idb-keyval';

export class FileOps {
  // file
  static async getFile(wsPath) {
    return idb.get(wsPath);
  }

  static async deleteFile(wsPath) {
    const file = await this.getFile(wsPath);
    if (!file) {
      throw new FileError(`File ${wsPath} does not exists`);
    }
    return idb.del(wsPath);
  }

  static async renameFile(wsPath, newWsPath) {
    const file = await this.getFile(wsPath);

    if (!file) {
      throw new FileError(`File ${wsPath} does not exists`);
    }

    const newFile = await this.getFile(newWsPath);

    if (newFile) {
      throw new FileError(`File ${newWsPath} already exists`);
    }

    await this.deleteFile(wsPath);
    return this.createFile(newWsPath, file);
  }

  static async createFile(wsPath, payload) {
    if (await this.getFile(wsPath)) {
      throw new FileError(`File ${wsPath} already exists`);
    }
    return idb.set(wsPath, payload);
  }

  static async updateFile(wsPath, payload) {
    if (!(await this.getFile(wsPath))) {
      throw new FileError(`File ${wsPath} does not exists`);
    }
    return idb.set(wsPath, payload);
  }

  static async listFiles(wsName) {
    const keys = (await idb.keys()) || [];

    return keys.filter((k) => k.startsWith(wsName + ':'));
  }
}

export class FileError extends Error {
  constructor(message) {
    super('FileError: ' + message);
    this.name = 'FileError';
  }
}
export class FilePermissionError extends Error {
  constructor(message) {
    super('FilePermissionError: ' + message);
    this.name = 'FilePermissionError';
  }
}
