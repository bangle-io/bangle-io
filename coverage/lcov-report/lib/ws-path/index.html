
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for lib/ws-path</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> lib/ws-path</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">75.72% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>234/309</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">67.32% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>68/101</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">81.48% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>66/81</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">75.75% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>225/297</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="helpers.ts"><a href="helpers.ts.html">helpers.ts</a></td>
	<td data-value="56.58" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 56%"></div><div class="cover-empty" style="width: 44%"></div></div>
	</td>
	<td data-value="56.58" class="pct medium">56.58%</td>
	<td data-value="129" class="abs medium">73/129</td>
	<td data-value="45.23" class="pct low">45.23%</td>
	<td data-value="42" class="abs low">19/42</td>
	<td data-value="61.53" class="pct medium">61.53%</td>
	<td data-value="26" class="abs medium">16/26</td>
	<td data-value="55.73" class="pct medium">55.73%</td>
	<td data-value="122" class="abs medium">68/122</td>
	</tr>

<tr>
	<td class="file high" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="location-helpers.ts"><a href="location-helpers.ts.html">location-helpers.ts</a></td>
	<td data-value="89.61" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 89%"></div><div class="cover-empty" style="width: 11%"></div></div>
	</td>
	<td data-value="89.61" class="pct high">89.61%</td>
	<td data-value="77" class="abs high">69/77</td>
	<td data-value="81.25" class="pct high">81.25%</td>
	<td data-value="16" class="abs high">13/16</td>
	<td data-value="84.61" class="pct high">84.61%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="90.41" class="pct high">90.41%</td>
	<td data-value="73" class="abs high">66/73</td>
	</tr>

<tr>
	<td class="file high" data-value="opened-ws-paths.ts"><a href="opened-ws-paths.ts.html">opened-ws-paths.ts</a></td>
	<td data-value="89" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 89%"></div><div class="cover-empty" style="width: 11%"></div></div>
	</td>
	<td data-value="89" class="pct high">89%</td>
	<td data-value="100" class="abs high">89/100</td>
	<td data-value="83.72" class="pct high">83.72%</td>
	<td data-value="43" class="abs high">36/43</td>
	<td data-value="92.85" class="pct high">92.85%</td>
	<td data-value="42" class="abs high">39/42</td>
	<td data-value="88.88" class="pct high">88.88%</td>
	<td data-value="99" class="abs high">88/99</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-12-02T21:35:16.931Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    