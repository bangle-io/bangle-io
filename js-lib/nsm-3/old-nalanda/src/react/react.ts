import { useDebugValue, useRef, useState } from 'react';
import useSyncExternalStoreExports from 'use-sync-external-store/shim';

import type { Slice } from '../vanilla/slice';
import type { Store } from '../vanilla/store';
import type { AnySlice } from '../vanilla/types';

const { useSyncExternalStore } = useSyncExternalStoreExports;

type InferData<T> = T extends Slice<any, infer TData, any> ? TData : never;

interface ReactAdapter<TSnapshot> {
  subscribe: (onStoreChange: () => void) => () => void;
  getSnapshot: () => TSnapshot;
}

export function createUseTrackSliceHook<TAllSliceName extends string = any>(
  store: Store<TAllSliceName>,
) {
  function useSlice<TSlice extends AnySlice>(sl: TSlice): InferData<TSlice> {
    const ref = useRef<InferData<TSlice>>();

    const [adapter] = useState(() => {
      const adapter: ReactAdapter<InferData<TSlice>> = {
        subscribe: (onStoreChange) => {
          const sliceEffect = store.effect((effectStore) => {
            const selectedData = sl.track(effectStore);
            ref.current = selectedData;

            onStoreChange();

            queueMicrotask(() => {
              if (effectStore._runInstance?._addTrackedCount === 0) {
                console.warn(
                  `You forgot to destructure/access the tracked field from ${sl.name}. This will not track any changes!.`,
                );
              }
            });
          });

          return () => {
            store.unregisterEffect(sliceEffect);
          };
        },
        getSnapshot: () => {
          return ref.current ?? sl.get(store.state);
        },
      };

      return adapter;
    });

    const data = useSyncExternalStore(adapter.subscribe, adapter.getSnapshot);

    useDebugValue(data);

    return data;
  }

  return useSlice;
}
