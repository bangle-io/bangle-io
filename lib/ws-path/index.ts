export * from './helpers';
export * from './opened-ws-paths';
