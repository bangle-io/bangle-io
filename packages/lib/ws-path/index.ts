export * from './helpers';
export * as locationHelpers from './location-helpers';
export * from './opened-ws-paths';
