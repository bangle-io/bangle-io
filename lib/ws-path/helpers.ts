import { BaseError } from '@bangle.io/base-error';
import { getLast } from '@bangle.io/mini-js-utils';
import type { WsName, WsPath } from '@bangle.io/shared-types';

export const DEFAULT_NOTE_EXTENSION = '.md';
export const VALID_NOTE_EXTENSIONS = [DEFAULT_NOTE_EXTENSION];

// works on any string
export function hasValidNoteExtension(str: string) {
  return VALID_NOTE_EXTENSIONS.some((ext) => str.endsWith(ext));
}

export function createWsName(wsName: string): WsName {
  validWsName(wsName);

  return wsName as WsName;
}

export function createWsPath(wsPath: string, validate = true): WsPath {
  if (validate) {
    validateWsPath(wsPath);
  }

  return wsPath as WsPath;
}

export function getExtension(str: string) {
  if (str.includes('/')) {
    str = str.slice(str.lastIndexOf('/') + 1);
  }

  const dotIndex = str.lastIndexOf('.');

  return dotIndex === -1 ? undefined : str.slice(dotIndex);
}

function conditionalSuffix(str: string, part: string) {
  if (str.endsWith(part)) {
    return str;
  }

  return str + part;
}

export function suffixWithNoteExtension(str: string) {
  return conditionalSuffix(str, DEFAULT_NOTE_EXTENSION);
}

export function removeExtension(str: string) {
  const dotIndex = str.lastIndexOf('.');

  return dotIndex === -1 ? str : str.slice(0, dotIndex);
}

export type MaybeWsPath = string | undefined;

/**
 * Types of paths
 * filePath - /a/b/c.md simple file path
 * wsPath - <wsName>:<filePath>
 * fsPath - /<wsName>/<filePath> - this is what is used internally by the fs module
 */
// TODO add test where wsPath has `//`
export function resolvePath(wsPath: string, skipValidation = false) {
  if (!skipValidation) {
    validateWsPath(wsPath);
    // TODO currently this only works for fileWsPaths
    validateFileWsPath(wsPath);
  }
  const [wsName, filePath] = splitWsPath(wsPath);
  const filePathSplitted = filePath.split('/');
  const fileName: string | undefined = getLast(filePathSplitted);

  if (typeof fileName !== 'string') {
    throw new Error('fileName undefined');
  }
  const dirPath = filePathSplitted
    .slice(0, filePathSplitted.length - 1)
    .filter(Boolean)
    .join('/');

  return {
    wsPath,
    wsName,
    filePath, // wsName:filePath
    dirPath, // wsName:dirPath/fileName
    fileName,
    fileNameWithoutExt: removeExtension(fileName),
  };
}

export function resolvePath2(wsPath: WsPath): {
  wsName: WsName;
  fileName: string;
  filePath: string;
  dirPath: string;
  fileNameWithoutExt: string;
} {
  const { wsName, fileName, filePath, dirPath, fileNameWithoutExt } =
    resolvePath(wsPath, true);

  return {
    wsName: createWsName(wsName),
    fileName,
    filePath,
    dirPath,
    fileNameWithoutExt,
  };
}

export class PathValidationError extends BaseError {}

export function isValidWsName(wsName: string | undefined): boolean {
  if (!wsName) {
    return false;
  }
  if (wsName.includes(':')) {
    return false;
  }

  return true;
}

export function validWsName(wsName: string) {
  if (wsName === '') {
    throw new PathValidationError({
      message: 'Invalid wsName "' + wsName + '" .',
    });
  }
  if (wsName.includes(':')) {
    throw new PathValidationError({
      message: 'Invalid characters in  "' + wsName + '" .',
    });
  }
}

export function isWsPath(wsPath: string) {
  if (!wsPath || typeof wsPath !== 'string') {
    return false;
  }

  if (wsPath.split('/').some((r) => r.length === 0)) {
    return false;
  }

  const [wsName, filePath, ...others] = wsPath.split(':');

  if (others.length > 0) {
    return false;
  }

  if (!wsName || !filePath) {
    return false;
  }

  return true;
}

export function isValidFileWsPath(wsPath: string) {
  if (!isWsPath(wsPath)) {
    return false;
  }

  const items = wsPath.split('/');

  if (items[items.length - 1]?.includes('.')) {
    return true;
  }

  return false;
}

export function validateWsPath(wsPath: string) {
  if (wsPath.split('/').some((r) => r.length === 0)) {
    throw new PathValidationError({ message: 'Invalid path ' + wsPath });
  }

  const [wsName, filePath, ...others] = wsPath.split(':');

  if (others.length > 0) {
    throw new PathValidationError({
      message: 'Semicolon not allowed file path',
    });
  }

  if (!isWsPath(wsPath)) {
    throw new PathValidationError({ message: 'Invalid wsPath ' + wsPath });
  }
}

// a file wsPath is workspace path to a file
export function validateFileWsPath(wsPath: string) {
  if (!isValidFileWsPath(wsPath)) {
    throw new PathValidationError({ message: 'Invalid wsPath ' + wsPath });
  }

  validateWsPath(wsPath);
}

// a note wsPath is any wsPath that is a note i.e. ends with VALID_NOTE_EXTENSIONS
export function validateNoteWsPath(wsPath: string) {
  validateFileWsPath(wsPath);

  if (!isValidNoteWsPath(wsPath)) {
    throw new PathValidationError({
      message: `Bangle.io support the following file extensions for notes: ${VALID_NOTE_EXTENSIONS.join(
        ', ',
      )}`,
    });
  }
}

export function isValidNoteWsPath(wsPath: string | undefined) {
  return wsPath ? hasValidNoteExtension(wsPath) : false;
}

export function sanitizeFilePath(filePath: string) {
  return filePath.replace(/[^\w\s-\.]/g, '');
}

/**
 * The local file paths are the paths supported by the file system.
 * For example ./my-file is a relative file.
 * @param {String} filePath - The file path read directly form the user input like an md file
 * @param {String} wsPath - the current file wsPath to resolve the relative path from
 * @returns {String} a valid wsPath to it
 */
export function parseLocalFilePath(filePath: string, wsPath: WsPath): WsPath {
  if (filePath.includes(':')) {
    throw new PathValidationError({
      message: 'Invalid character ":" in  "' + filePath + '" .',
    });
  }

  if (filePath.startsWith('./')) {
    filePath = filePath.slice(2);
  }
  const { wsName, dirPath } = resolvePath2(wsPath);
  let sampleDomain = 'https://bangle.io';

  if (dirPath) {
    sampleDomain += '/' + dirPath + '/';
  }
  let webPath = new URL(filePath, sampleDomain).pathname;

  if (webPath.startsWith('/')) {
    webPath = webPath.slice(1);
  }

  // need to decode uri as filesystems dont do encoding
  return filePathToWsPath2(wsName, decodeURIComponent(webPath));
}

export function parseLocalFilePath2(filePath: string, wsPath: WsPath): WsPath {
  return createWsPath(parseLocalFilePath(filePath, wsPath));
}

export const toFSPath = (wsPath: string) => {
  const { wsName, filePath } = resolvePath(wsPath);

  return [wsName, filePath].join('/');
};

export function fromFsPath(fsPath: string) {
  const [_wsName, ...f] = fsPath.split('/');

  if (!_wsName || _wsName.includes(':')) {
    return undefined;
  }

  return filePathToWsPath(_wsName, f.join('/'));
}

export function splitWsPath(wsPath: string): [string, string] {
  const [wsName, filePath] = wsPath.split(':');

  if (!wsName) {
    throw new PathValidationError({ message: 'Invalid wsName' });
  }

  if (!filePath) {
    throw new PathValidationError({ message: 'Invalid filePath' });
  }

  return [wsName, filePath];
}

export function updateFileName2(wsPath: WsPath, newFileName: string): WsPath {
  const { wsName, dirPath } = resolvePath2(wsPath);

  return filePathToWsPath2(wsName, `${dirPath}/${newFileName}`);
}

export function filePathToWsPath(wsName: string, filePath: string) {
  if (filePath.startsWith('/')) {
    filePath = filePath.slice(1);
  }

  return wsName + ':' + filePath;
}

export function filePathToWsPath2(wsName: WsName, filePath: string): WsPath {
  if (filePath.startsWith('/')) {
    filePath = filePath.slice(1);
  }

  return createWsPath(wsName + ':' + filePath);
}
