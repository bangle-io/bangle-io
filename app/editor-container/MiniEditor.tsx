import React, { useCallback, useState } from 'react';

import { useNsmStore } from '@bangle.io/bangle-store-context';
import { EditorDisplayType, MINI_EDITOR_INDEX } from '@bangle.io/constants';
import { vars } from '@bangle.io/css-vars';
import { Editor } from '@bangle.io/editor';
import {
  nsmSliceWorkspace,
  pushOpenedWsPaths,
} from '@bangle.io/nsm-slice-workspace';
import type { EternalVars, WsPath } from '@bangle.io/shared-types';
import { nsmPageSlice } from '@bangle.io/slice-page';
import {
  ArrowsExpand,
  Button,
  ChevronDownIcon,
  ChevronUpIcon,
  CloseIcon,
} from '@bangle.io/ui-components';
import { resolvePath } from '@bangle.io/ws-path';

export function MiniEditor({
  wsPath,
  eternalVars,
}: {
  wsPath: WsPath;
  eternalVars: EternalVars;
}) {
  const { fileNameWithoutExt } = resolvePath(wsPath, true);

  const nsmStore = useNsmStore([nsmPageSlice, nsmSliceWorkspace]);
  const [isMinimized, updateIsMinimized] = useState(false);

  const onClose = useCallback(() => {
    nsmStore.dispatch(
      pushOpenedWsPaths((oPaths) => oPaths.updateMiniEditorWsPath(undefined)),
    );
  }, [nsmStore]);

  const onExpand = useCallback(() => {
    nsmStore.dispatch(
      pushOpenedWsPaths((oPaths) => oPaths.updateMiniEditorWsPath(undefined)),
    );
    nsmStore.dispatch(
      pushOpenedWsPaths((oPaths) => {
        const result = oPaths.updatePrimaryWsPath(wsPath);

        return result;
      }),
    );
  }, [wsPath, nsmStore]);

  return (
    <div
      data-testid="editor-container_mini-editor-wrapper"
      // setting width and height are important otherwise
      // tippy confuses and causes issue with scroll
      className="fixed flex flex-col bottom-0 right-8 rounded drop-shadow-xl z-popup border-neutral"
      style={{
        width: vars.misc.miniEditorWidth,
        background: vars.misc.editorBg,
      }}
    >
      <div className="flex flex-row w-full px-2 py-1 text-sm justify-between border-b-1 border-colorNeutralBorder">
        <div
          className="font-semibold flex items-center flex-grow cursor-pointer truncate select-none"
          onClick={() => {
            updateIsMinimized((e) => !e);
          }}
        >
          <span className="truncate">{fileNameWithoutExt}</span>
        </div>
        <div className="flex flex-row">
          {isMinimized ? (
            <Button
              size="xs"
              variant="transparent"
              ariaLabel="Maximize"
              onPress={() => {
                updateIsMinimized((e) => !e);
              }}
              leftIcon={<ChevronUpIcon />}
            />
          ) : (
            <Button
              size="xs"
              variant="transparent"
              ariaLabel="Minimize"
              leftIcon={<ChevronDownIcon />}
              onPress={() => {
                updateIsMinimized((e) => !e);
              }}
            />
          )}
          <Button
            variant="transparent"
            size="xs"
            onPress={onExpand}
            ariaLabel={'Expand to full screen'}
            leftIcon={<ArrowsExpand />}
          />
          <Button
            variant="transparent"
            size="xs"
            onPress={onClose}
            ariaLabel={'Close'}
            leftIcon={<CloseIcon />}
          />
        </div>
      </div>
      {isMinimized ? null : (
        <div
          className="px-2 overflow-y-auto pl-6 B-editor-container_mini-editor"
          style={{
            height: 'min(50vh, 600px)',
          }}
        >
          <Editor
            editorDisplayType={EditorDisplayType.Floating}
            editorId={MINI_EDITOR_INDEX}
            eternalVars={eternalVars}
            wsPath={wsPath}
          />
        </div>
      )}
    </div>
  );
}
