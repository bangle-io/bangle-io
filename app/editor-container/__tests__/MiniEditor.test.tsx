/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import {
  createBasicTestStore,
  setupMockWorkspaceWithNotes,
  TestStoreProvider,
} from '@bangle.io/test-utils';
import { resolvePath } from '@bangle.io/ws-path';

import { MiniEditor } from '../MiniEditor';

const DOC_CONTENT = `hello mars`;

describe('MiniEditor', () => {
  const setup = async () => {
    let { store, getActionNames } = createBasicTestStore({
      extensions: [],
      useEditorCoreExtension: true,
      useUISlice: true,
      useEditorManagerSlice: true,
    });

    const wsPath = 'my-ws:test-dir/magic.md';

    const { wsName } = resolvePath(wsPath);

    await setupMockWorkspaceWithNotes(store, wsName, [[wsPath, DOC_CONTENT]]);

    const { container, getByLabelText } = render(
      <div>
        <TestStoreProvider bangleStore={store} bangleStoreChanged={0}>
          <MiniEditor wsPath={wsPath} />
        </TestStoreProvider>
      </div>,
    );

    await waitFor(() => {
      expect(container.innerHTML).toContain(DOC_CONTENT);
    });

    return { container, getByLabelText, getActionNames };
  };

  test('works', async () => {
    const { container } = await setup();
    expect(container.innerHTML).toContain('hello mars');
    expect(container).toMatchSnapshot();
  });

  test('minimize works', async () => {
    const { container, getByLabelText } = await setup();

    expect(container.innerHTML).toContain('hello mars');

    fireEvent.click(getByLabelText('Minimize'));

    expect(container.innerHTML).not.toContain('hello mars');
  });
});
