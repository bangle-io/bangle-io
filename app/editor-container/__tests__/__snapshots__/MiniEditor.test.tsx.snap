// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniEditor works 1`] = `
<div>
  <div>
    <div
      class="fixed flex flex-col bottom-0 right-8 rounded drop-shadow-xl z-popup border-neutral"
      data-testid="editor-container_mini-editor-wrapper"
    >
      <div
        class="flex flex-row w-full px-2 py-1 text-sm justify-between border-b-1 border-colorNeutralBorder"
      >
        <div
          class="font-semibold flex items-center flex-grow cursor-pointer truncate select-none"
        >
          <span
            class="truncate"
          >
            magic
          </span>
        </div>
        <div
          class="flex flex-row"
        >
          <button
            aria-label="Minimize"
            class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
            style="background-color: transparent;"
            type="button"
          >
            <span
              class="flex flex-grow-1 overflow-hidden "
              style="justify-content: center;"
            >
              <span>
                <svg
                  class=" w-4 h-4  "
                  fill="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.343 7.757L4.93 9.172 12 16.242l7.071-7.07-1.414-1.415L12 13.414 6.343 7.757z"
                  />
                </svg>
              </span>
            </span>
          </button>
          <button
            aria-label="Expand to full screen"
            class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
            style="background-color: transparent;"
            type="button"
          >
            <span
              class="flex flex-grow-1 overflow-hidden "
              style="justify-content: center;"
            >
              <span>
                <svg
                  class="w-4 h-4  "
                  fill="none"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 8V4m0 0h4M4 4l5 5m11-1V4m0 0h-4m4 0l-5 5M4 16v4m0 0h4m-4 0l5-5m11 5l-5-5m5 5v-4m0 4h-4"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </span>
            </span>
          </button>
          <button
            aria-label="Close"
            class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
            style="background-color: transparent;"
            type="button"
          >
            <span
              class="flex flex-grow-1 overflow-hidden "
              style="justify-content: center;"
            >
              <span>
                <svg
                  class=" w-4 h-4  "
                  fill="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        class="px-2 overflow-y-auto pl-6 B-editor-container_mini-editor"
      >
        <div
          class="B-editor_display-popup "
        >
          <div
            class="ProseMirror bangle-editor"
            contenteditable="true"
            translate="no"
          >
            <p>
              hello mars
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
